File: maven-plugin-v2/src/main/java/CreateDependenciesGenerator.java
import model.*;
import org.apache.maven.project.MavenProject;
import org.apache.maven.model.Dependency;
import org.apache.maven.plugin.logging.Log;
import java.io.File;
import java.util.*;

/**
 * Generates CreateDependencies compatible results for Nx integration
 */
public class CreateDependenciesGenerator {

    /**
     * Generate CreateDependencies results from Maven project list
     * Returns: List<RawProjectGraphDependency>
     */
    public static List<RawProjectGraphDependency> generateCreateDependencies(List<MavenProject> projects, File workspaceRoot) {
        return generateCreateDependencies(projects, workspaceRoot, null, false);
    }

    /**
     * Generate CreateDependencies results with logging support
     */
    public static List<RawProjectGraphDependency> generateCreateDependencies(List<MavenProject> projects, File workspaceRoot, Log log, boolean verbose) {
        List<RawProjectGraphDependency> dependencies = new ArrayList<>();

        // Extract workspace projects
        Map<String, String> artifactToProject = buildArtifactMapping(projects);

        if (verbose && log != null) {
            log.info("Built artifact mapping for " + artifactToProject.size() + " workspace projects");
        }

        int staticDeps = 0;
        int implicitDeps = 0;

        for (int i = 0; i < projects.size(); i++) {
            MavenProject project = projects.get(i);
            String source = NxPathUtils.getProjectName(project);
            String sourceFile = NxPathUtils.getRelativePomPath(project, workspaceRoot);

            log.info("Dependency analysis progress: " + (i + 1) + "/" + projects.size() + " projects");

            // 1. Static dependencies - direct Maven dependencies
            int prevStatic = staticDeps;
            staticDeps += addStaticDependencies(dependencies, project, artifactToProject, sourceFile);

            // 2. Implicit dependencies - parent/module relationships
            int prevImplicit = implicitDeps;
            implicitDeps += addImplicitDependencies(dependencies, project, projects, source);

            if (verbose && log != null && projects.size() <= 20) {
                int newStatic = staticDeps - prevStatic;
                int newImplicit = implicitDeps - prevImplicit;
                if (newStatic > 0 || newImplicit > 0) {
                    log.info("Project " + source + ": " + newStatic + " static, " + newImplicit + " implicit dependencies");
                }
            }
        }

        if (verbose && log != null) {
            log.info("Dependency analysis complete: " + staticDeps + " static, " + implicitDeps + " implicit dependencies");
        }

        return dependencies;
    }

    /**
     * Build mapping from artifactId to project name for workspace projects
     */
    private static Map<String, String> buildArtifactMapping(List<MavenProject> projects) {
        Map<String, String> mapping = new HashMap<>();
        for (MavenProject project : projects) {
            String projectName = NxPathUtils.getProjectName(project);
            if (project.getGroupId() != null && project.getArtifactId() != null) {
                String key = project.getGroupId() + ":" + project.getArtifactId();
                mapping.put(key, projectName);
            }
        }
        return mapping;
    }

    /**
     * Add static dependencies (explicit Maven dependencies between workspace projects)
     * Returns: number of dependencies added
     */
    private static int addStaticDependencies(List<RawProjectGraphDependency> dependencies,
                                              MavenProject project,
                                              Map<String, String> artifactToProject,
                                              String sourceFile) {
        String source = NxPathUtils.getProjectName(project);
        int count = 0;

        // Check both declared dependencies and resolved artifacts
        if (project.getDependencies() != null) {
            for (Dependency dep : project.getDependencies()) {
                if (dep.getGroupId() != null && dep.getArtifactId() != null) {
                    String depKey = dep.getGroupId() + ":" + dep.getArtifactId();

                    // Check if this dependency refers to another project in workspace
                    String target = artifactToProject.get(depKey);
                    if (target != null && !target.equals(source)) {
                        RawProjectGraphDependency dependency = new RawProjectGraphDependency(
                            source, target, RawProjectGraphDependency.DependencyType.STATIC, sourceFile);
                        dependencies.add(dependency);
                        count++;
                    }
                }
            }
        }

        return count;
    }

    /**
     * Add implicit dependencies (parent-child module relationships)
     * Returns: number of dependencies added
     */
    private static int addImplicitDependencies(List<RawProjectGraphDependency> dependencies,
                                                MavenProject project,
                                                List<MavenProject> allProjects,
                                                String source) {
        // Find parent-child relationships
        File projectDir = project.getBasedir();
        int count = 0;

        for (MavenProject otherProject : allProjects) {
            File otherDir = otherProject.getBasedir();
            String target = NxPathUtils.getProjectName(otherProject);

            if (!source.equals(target)) {
                // Check if one is parent of the other
                if (NxPathUtils.isParentChildRelation(projectDir, otherDir)) {
                    RawProjectGraphDependency dependency = new RawProjectGraphDependency(
                        source, target, RawProjectGraphDependency.DependencyType.IMPLICIT);
                    dependencies.add(dependency);
                    count++;
                }
            }
        }

        return count;
    }


    /**
     * Check if a dependency is a test-scoped dependency
     */
    private static boolean isTestDependency(Dependency dep) {
        return "test".equals(dep.getScope());
    }

    /**
     * Check if a dependency is optional
     */
    private static boolean isOptionalDependency(Dependency dep) {
        return dep.isOptional();
    }
}

File: maven-plugin/src/main/kotlin/CreateDependenciesGenerator.kt
import model.RawProjectGraphDependency
import org.apache.maven.project.MavenProject
import org.apache.maven.model.Dependency
import org.apache.maven.plugin.logging.Log
import java.io.File

/**
 * Generates CreateDependencies compatible results for Nx integration
 */
object CreateDependenciesGenerator {

    /**
     * Generate CreateDependencies results from Maven project list
     * Returns: List<RawProjectGraphDependency>
     */
    fun generateCreateDependencies(projects: List<MavenProject>, workspaceRoot: File): List<RawProjectGraphDependency> {
        return generateCreateDependencies(projects, workspaceRoot, null, false)
    }

    /**
     * Generate CreateDependencies results with logging support
     */
    fun generateCreateDependencies(
        projects: List<MavenProject>, 
        workspaceRoot: File, 
        log: Log?, 
        verbose: Boolean
    ): List<RawProjectGraphDependency> {
        val dependencies = mutableListOf<RawProjectGraphDependency>()

        // Extract workspace projects
        val artifactToProject = buildArtifactMapping(projects)

        if (verbose && log != null) {
            log.info("Built artifact mapping for ${artifactToProject.size} workspace projects")
        }

        var staticDeps = 0

        for (i in projects.indices) {
            val project = projects[i]
            val source = MavenUtils.formatProjectKey(project)
            val sourceFile = NxPathUtils.getRelativePomPath(project, workspaceRoot)

            // Show progress every 100 projects to reduce log spam
            if (i % 100 == 0 || i == projects.size - 1) {
                log?.info("Dependency analysis progress: ${i + 1}/${projects.size} projects")
            }

            // 1. Static dependencies - direct Maven dependencies
            val prevStatic = staticDeps
            staticDeps += addStaticDependencies(dependencies, project, artifactToProject, sourceFile)

            // 2. Skip implicit dependencies for better performance
            // They're rarely needed and cause O(n²) complexity

            if (verbose && log != null && projects.size <= 20) {
                val newStatic = staticDeps - prevStatic
                if (newStatic > 0) {
                    log.info("Project $source: $newStatic static dependencies")
                }
            }
        }

        if (verbose && log != null) {
            log.info("Dependency analysis complete: $staticDeps static dependencies")
        }

        return dependencies
    }

    /**
     * Build mapping from artifactId to project name for workspace projects
     */
    private fun buildArtifactMapping(projects: List<MavenProject>): Map<String, String> {
        val mapping = mutableMapOf<String, String>()
        for (project in projects) {
            if (project.groupId != null && project.artifactId != null) {
                val key = MavenUtils.formatProjectKey(project)
                val projectName = MavenUtils.formatProjectKey(project)
                mapping[key] = projectName
            } else {
                System.err.println("Warning: Skipping project with null groupId or artifactId: $project")
            }
        }
        return mapping
    }

    /**
     * Add static dependencies (explicit Maven dependencies between workspace projects)
     * Returns: number of dependencies added
     */
    private fun addStaticDependencies(
        dependencies: MutableList<RawProjectGraphDependency>,
        project: MavenProject,
        artifactToProject: Map<String, String>,
        sourceFile: String
    ): Int {
        val source = MavenUtils.formatProjectKey(project)
        var count = 0

        // Check both declared dependencies and resolved artifacts
        project.dependencies?.let { deps ->
            for (dep in deps) {
                if (dep.groupId != null && dep.artifactId != null) {
                    val depKey = "${dep.groupId}:${dep.artifactId}"

                    // Check if this dependency refers to another project in workspace
                    val target = artifactToProject[depKey]
                    if (target != null && target != source) {
                        val dependency = RawProjectGraphDependency(
                            source, target, RawProjectGraphDependency.DependencyType.STATIC, sourceFile
                        )
                        dependencies.add(dependency)
                        count++
                    }
                }
            }
        }

        return count
    }
}